README.md
# RAVA Python Driver

The [RAVA Python Driver](https://github.com/gabrielguerrer/rng_rava_driver_py) 
implements the code for communicating with an 
[RAVA Device](https://github.com/gabrielguerrer/rng_rava) running the 
[RAVA Firmware](https://github.com/gabrielguerrer/rng_rava_firmware). 
The computer running the driver assumes the role of the leader device, sending 
command requests and reading the data replies.

The RAVA_RNG class enables the request of pulse counts, random bits, random 
bytes, and random numbers (integers and floats). Additionally, it establishes 
the circuit's basic functionality encompassing key modules such as EEPROM, PWM, 
heath tests, peripherals, and interfaces.

The RAVA_RNG_AIO class offers the same functionality as RAVA_RNG but within an 
asynchronous framework.

The RAVA_RNG_LED class implements the code for controlling the LED and LAMP 
modules within the RAVA circuit. It allows users to adjust the color and the 
intensity of the attached LED. Users can also activate the LAMP mode and 
retrieve statistical information on its operation.

For a deeper understanding of how the driver operates, please refer to the 
documentation provided within the source files and the provided examples.


## Installation

The driver code is available as the 
[rng_rava](https://pypi.org/project/rng_rava/) PyPI package. To install it, run:

```
pip install rng_rava
```

Requirements:
 * [pyserial](https://github.com/pyserial/pyserial) 

## Usage

```
import rng_rava as rava

# Find the serial number of the attached RAVA devices
rava_sns = rava.find_rava_sns()

# Create a RNG instance and connect to the first device
rng = rava.RAVA_RNG()
rng.connect(serial_number=rava_sns[0])

'''
The default PWM and RNG configuration parameters are stored in the EEPROM memory
and can be accessed with rng.get_eeprom_pwm() and rng.get_eeprom_rng(). If
desired, users can modify the default values using the respective snd_
functions. Additionally, it is possible to make non-permanent configuration
changes using the following commands:
'''

# Configure PWM
rng.snd_pwm_setup(freq_id=rava.D_PWM_FREQ['50_KHZ'], duty=20)

# Configure RNG
rng.snd_rng_setup(sampling_interval_us=10)

'''
Next, the generation of various random data types is demonstrated.
'''

# Measure 100 pulse counts
pc_a, pc_b = rng.get_rng_pulse_counts(n_counts=100)

# Generate a random bit XORing both channels
bit = rng.get_rng_bits(bit_source_id=rava.D_RNG_BIT_SRC['AB_XOR'])

# Generate 100 random bytes en each channel without post-processing
# Output as numpy array
bytes_a, bytes_b = rng.get_rng_bytes(n_bytes=100,
                                     postproc_id=rava.D_RNG_POSTPROC['NONE'],
                                     list_output=True)

# Generate 100 8-bit integers between 0 and 99
ints8 = rng.get_rng_int8s(n_ints=100, int_delta=100)

# Generate 100 16-bit integers between 0 and 9999
ints16 = rng.get_rng_int16s(n_ints=100, int_delta=10000)

# Generate 100 32-bit floats ranging between 0 and 1
floats = rng.get_rng_floats(n_floats=100)
```

## Associated projects

- [RAVA Device](https://github.com/gabrielguerrer/rng_rava)
- [RAVA Firmware](https://github.com/gabrielguerrer/rng_rava_firmware)
- RAVA Python Diagnostics Tool

## Contact

gabrielguerrer [at] gmail [dot] com

![RAVA logo](https://github.com/gabrielguerrer/rng_rava/blob/main/images/rng_rava_logo.png)

CHANGELOG.md
## v1.0.1

* Removed numpy dependency
* Changed maximum line length from 80 to 120. This change does not apply to the code's documentation
* Using "not in" and "is not None"
* Correcting firmware version in eeprom_firmware
* Adding callbacks
* Setting logger name to 'rava'
* Changed health startup results format
* Including hardware float generation (and moved software double to examples)

examples/led_test.py
'''
This example showcases LED capabilities.

This example code is in the public domain.
Author: Gabriel Guerrer
'''

import time
import rng_rava as rava

# Find RAVA device and connect
rngled = rava.RAVA_RNG_LED()
dev_sns = rava.find_rava_sns()
if len(dev_sns):
    rngled.connect(serial_number=dev_sns[0])
else:
    rava.lg.error('No device found')
    exit()

# Set color as blue (intensity=0)
rngled.snd_led_color(color_hue=rava.D_LED_COLOR['BLUE'], intensity=0)

# Fade to full intensity
rngled.snd_led_intensity_fade(intensity_tgt=255, duration_ms=2000)
time.sleep(3)

# Oscilate colors
rngled.snd_led_color_oscillate(n_cycles=3, duration_ms=4000)
time.sleep(5)

# Oscilate intensities
for i in range(2):
    rngled.snd_led_intensity_fade(intensity_tgt=10, duration_ms=2000)
    time.sleep(2)
    rngled.snd_led_intensity_fade(intensity_tgt=255, duration_ms=2000)
    time.sleep(2)
time.sleep(1)

# Fade color to red
rngled.snd_led_color_fade(color_hue_tgt=rava.D_LED_COLOR['RED'], duration_ms=2000)
time.sleep(3)

# Turn off lights
rngled.snd_led_intensity(intensity=0)

# Close device
rngled.close()

examples/rng_example.py
'''
RAVA RNG usage example.

This example code is in the public domain.
Author: Gabriel Guerrer
'''

import rng_rava as rava

# Find the serial number of the attached RAVA devices
rava_sns = rava.find_rava_sns()

# Create a RNG instance and connect to the first device
rng = rava.RAVA_RNG()
rng.connect(serial_number=rava_sns[0])

'''
The default PWM and RNG configuration parameters are stored in the EEPROM memory
and can be accessed with rng.get_eeprom_pwm() and rng.get_eeprom_rng(). If
desired, users can modify the default values using the respective snd_
functions. Additionally, it is possible to make non-permanent configuration
changes using the following commands:
'''

# Configure PWM
rng.snd_pwm_setup(freq_id=rava.D_PWM_FREQ['50_KHZ'], duty=20)

# Configure RNG
rng.snd_rng_setup(sampling_interval_us=10)

'''
Next, the generation of various random data types is demonstrated.
'''

# Measure 100 pulse counts
pc_a, pc_b = rng.get_rng_pulse_counts(n_counts=100)

# Generate a random bit XORing both channels
bit = rng.get_rng_bits(bit_source_id=rava.D_RNG_BIT_SRC['AB_XOR'])

# Generate 100 random bytes en each channel without post-processing
# Output as numpy array
bytes_a, bytes_b = rng.get_rng_bytes(n_bytes=100,
                                     postproc_id=rava.D_RNG_POSTPROC['NONE'],
                                     list_output=True)

# Generate 100 8-bit integers between 0 and 99
ints8 = rng.get_rng_int8s(n_ints=100, int_delta=100)

# Generate 100 16-bit integers between 0 and 9999
ints16 = rng.get_rng_int16s(n_ints=100, int_delta=10000)

# Generate 100 32-bit floats ranging between 0 and 1
floats = rng.get_rng_floats(n_floats=100)

src/rng_rava/rava_defs.py
"""
Copyright (c) 2023 Gabriel Guerrer

Distributed under the MIT license - See LICENSE for details
"""

"""
Definitions and variables used by the RAVA modules.
"""

# RAVA_LOG_LEVEL = 10 # logging.DEBUG
RAVA_LOG_LEVEL = 20 # logging.INFO
LOG_FILL = 17 * ' '

RAVA_USB_VID = 0x1209 # https://pid.codes
RAVA_USB_PID = 0x4884

COMM_MSG_START = b'$'

COMM_MSG_LEN = 8

PERIPH_PORTS = 5

GET_TIMEOUT_S = 3.

SERIAL_LISTEN_LOOP_INTERVAL_S = 0.02 # 20 ms, 50 Hz readings

RNG_BYTE_STREAM_MAX_INTERVAL_MS = 4194

D_DEV_COMM = {
    'DEVICE_SERIAL_NUMBER':1,
    'DEVICE_TEMPERATURE':2,
    'DEVICE_FREE_RAM':3,
    'DEVICE_REBOOT':4,
    'DEVICE_DEBUG':5,

    'EEPROM_RESET_TO_DEFAULT':10,
    'EEPROM_DEVICE':11,
    'EEPROM_FIRMWARE':12,
    'EEPROM_PWM':13,
    'EEPROM_RNG':14,
    'EEPROM_LED':15,
    'EEPROM_LAMP':16,

    'PWM_SETUP':30,

    'RNG_SETUP':40,
    'RNG_PULSE_COUNTS':41,
    'RNG_BITS':42,
    'RNG_BYTES':43,
    'RNG_TIMING_DEBUG_D1':44,

    'RNG_INT8S':50,
    'RNG_INT16S':51,
    'RNG_FLOATS':52,

    'RNG_STREAM_START':60,
    'RNG_STREAM_STOP':61,
    'RNG_STREAM_BYTES':62,
    'RNG_STREAM_STATUS':63,

    'HEALTH_STARTUP_RUN':70,
    'HEALTH_STARTUP_RESULTS':71,
    'HEALTH_CONTINUOUS_ERRORS':72,

    'LED_COLOR':80,
    'LED_COLOR_FADE':81,
    'LED_COLOR_OSCILLATE':82,
    'LED_INTENSITY':83,
    'LED_INTENSITY_FADE':84,
    'LED_FADE_STOP':85,
    'LED_STATUS':86,

    'LAMP_MODE':90,
    'LAMP_STATISTICS':91,
    'LAMP_DEBUG':92,

    'PERIPH_MODE':100,
    'PERIPH_READ':101,
    'PERIPH_WRITE':102,
    'PERIPH_PULSE':103,

    'PERIPH_D1_TRIGGER_INPUT':110,
    'PERIPH_D1_COMPARATOR':111,
    'PERIPH_D1_DELAY_US_TEST':112,
    'PERIPH_D2_TIMER3_INPUT_CAPTURE':113,
    'PERIPH_D3_TIMER3_TRIGGER_OUTPUT':114,
    'PERIPH_D3_TIMER3_PWM':115,
    'PERIPH_D4_PIN_CHANGE':116,
    'PERIPH_D5_ADC':117,

    'INTERFACE_DS18B20':130
    }
D_DEV_COMM_INV = {v: k for k, v in D_DEV_COMM.items()}

D_PWM_FREQ = {
    '30_KHZ':1,
    '40_KHZ':2,
    '50_KHZ':3,
    '60_KHZ':4,
    '75_KHZ':5
    }
D_PWM_FREQ_INV = {v: k for k, v in D_PWM_FREQ.items()}

D_RNG_BIT_SRC = {
    'AB':1,
    'A':2,
    'B':3,
    'AB_XOR':4,
    'AB_RND':5
}
D_RNG_BIT_SRC_INV = {v: k for k, v in D_RNG_BIT_SRC.items()}

D_RNG_POSTPROC = {
    'NONE':0,
    'XOR':1,
    'XOR_DICHTL':2,
    'VON_NEUMANN':3
    }
D_RNG_POSTPROC_INV = {v: k for k, v in D_RNG_POSTPROC.items()}

D_PERIPH_MODES = {
    'INPUT':0,
    'OUTPUT':1
}
D_PERIPH_MODES_INV = {v: k for k, v in D_PERIPH_MODES.items()}

D_LED_COLOR = {
    'RED':0,
    'ORANGE':16,
    'YELLOW':32,
    'GREEN':96,
    'CYAN':128,
    'BLUE':160,
    'PURPLE':192,
    'PINK':224
    }
D_LED_COLOR_INV = {v: k for k, v in D_LED_COLOR.items()}

examples/rng_bytes_to_file.py
'''
This example showcases the generation of a binary file containing random bytes.

This example code is in the public domain.
Author: Gabriel Guerrer
'''

import rng_rava as rava

# Variables
FILE_OUTPUT = 'random.bin'
N_BYTES = 1000000 # 1MB

# Find RAVA device and connect
rng = rava.RAVA_RNG()
dev_sns = rava.find_rava_sns()
if len(dev_sns):
    rng.connect(serial_number=dev_sns[0])
else:
    rava.lg.error('No device found')
    exit()

# Open file
with open(FILE_OUTPUT, mode='bw') as f:

    # Generate bytes
    bytes_a, bytes_b = rng.get_rng_bytes(n_bytes=N_BYTES//2,
                                    postproc_id=rava.D_RNG_POSTPROC['NONE'],
                                    list_output=False,
                                    timeout=100)

    # Write to file
    f.write(bytes_a)
    f.write(bytes_b)

# Close device
rng.close()

src/rng_rava/rava_rng_aio.py
"""
Copyright (c) 2023 Gabriel Guerrer

Distributed under the MIT license - See LICENSE for details
"""

"""
The RAVA_RNG_AIO class offers the same functionality as RAVA_RNG but within an
asynchronous framework.
"""

import asyncio

from rng_rava.rava_defs import *
from rng_rava.rava_rng import *


class RAVA_RNG_AIO(RAVA_RNG):

    def __init__(self):
        super().__init__(dev_name='RAVA_RNG_AIO')
        self.queue_type = asyncio.Queue

        # Variables
        self.serial_listen_task = None


    async def connect(self, serial_number):
        # Debug
        lg.debug('> {} CONNECT'.format(self.dev_name))

        # Find serial port
        port = find_rava_port(serial_number)
        if port is None:
            lg.error('{} Connect: No device found with SN {}'.format(self.dev_name, serial_number))
            return False

        # Open serial connection
        if not self.open_serial(port):
            return False

        # Reset serial data queues
        self.init_queue_data()

        # Start listening for serial commands
        self.serial_listen_task = asyncio.create_task(self.loop_serial_listen())

        # Stop any active RNG byte stream
        self.snd_rng_byte_stream_stop()

        # Request firmware info
        self.dev_firmware_dict = await self.get_eeprom_firmware()

        # Print connection info
        lg.info('{} Connect: Success'
                '\n{}  {}, Firmware v{}, SN={}, at {}'
                .format(self.dev_name, LOG_FILL,
                        self.dev_usb_name, self.dev_firmware_dict['version'], self.dev_serial_number, self.serial.port))

        # Request Health startup info
        if self.dev_firmware_dict['health_startup_enabled']:
            test_success, test_vars = await self.get_health_startup_results()

            # Print test info
            print_health_startup_results(test_success, test_vars)

            # Error? Users have then a limited command variety (see Firmware)
            if not test_success:
                lg.error('{} Connect: Startup tests failed'.format(self.dev_name))
                return False

        return True


    def put_queue_data(self, comm, value, comm_ext_id=0):
        # Check key
        if comm not in self.serial_data:
            lg.error('{} Data: Unknown comm {}'.format(self.dev_name, comm))
            return False

        # Extended key
        if comm_ext_id:
            comm_ext = '{}_{}'.format(comm, comm_ext_id)

            # Queue exists?
            if comm_ext not in self.serial_data:
                self.serial_data[comm_ext] = self.queue_type()
        else:
            comm_ext = comm

        # Try writing to queue
        try:
            self.serial_data[comm_ext].put_nowait(value)
            return True

        except asyncio.QueueFull:
            lg.error('{} Data: {} Queue full'.format(self.dev_name, comm_ext))
            return False


    async def get_queue_data(self, comm, comm_ext_id=0):
        # Check key
        if comm not in self.serial_data:
            lg.error('{} Data: Unknown comm {}'.format(self.dev_name, comm))
            return None

        # Extended key
        if comm_ext_id:
            comm_ext = '{}_{}'.format(comm, comm_ext_id)

            # Queue exists?
            if comm_ext not in self.serial_data:
                self.serial_data[comm_ext] = self.queue_type()
        else:
            comm_ext = comm

        # Read asyncio queue
        return await self.serial_data[comm_ext].get()


    async def loop_serial_listen(self):
        try:
            # Debug
            lg.debug('> {} SERIAL LISTEN LOOP'.format(self.dev_name))

            # Loop while connected
            while self.serial_connected.is_set():

                # Command available?
                if self.inwaiting_serial():

                    # Starts with $?
                    if self.read_serial(1) == COMM_MSG_START:
                        comm_msg = self.read_serial(COMM_MSG_LEN-1)
                        comm_id = comm_msg[0]
                        comm_data = comm_msg[1:]

                        # Known command id?
                        if comm_id in D_DEV_COMM_INV:
                            # Debug
                            lg.debug('> COMM RCV {}'.format([D_DEV_COMM_INV[comm_id], *[c for c in comm_data]]))

                            # Process Command
                            self.process_serial_comm(comm_id, comm_data)

                        else:
                            lg.warning('{} Serial Listen Loop: Unknown command_id {}'.format(self.dev_name, comm_id))

                    else:
                        lg.warning('{} Serial Listen Loop: Commands must start with {}'.format(self.dev_name, COMM_MSG_START))

                # The non-blocking method is prefered for finishing the thread
                # when closing the device
                else:
                    await asyncio.sleep(SERIAL_LISTEN_LOOP_INTERVAL_S)

        except Exception as err:
            lg.error('{} Serial Listen Loop: Error'
                     '\n{}  {} - {}'
                     .format(self.dev_name, LOG_FILL, type(err).__name__, err))

            # Close device
            self.close()


    #####################
    ## DEVICE

    async def get_device_serial_number(self):
        comm = 'DEVICE_SERIAL_NUMBER'
        if self.snd_rava_msg(comm):
            return await self.get_queue_data(comm)


    async def get_device_temperature(self):
        comm = 'DEVICE_TEMPERATURE'
        if self.snd_rava_msg(comm):
            return await self.get_queue_data(comm)


    async def get_device_free_ram(self):
        comm = 'DEVICE_FREE_RAM'
        if self.snd_rava_msg(comm):
            return await self.get_queue_data(comm)



    #####################
    ## EEPROM

    async def get_eeprom_device(self):
        comm = 'EEPROM_DEVICE'
        rava_send = True
        if self.snd_rava_msg(comm, [rava_send], 'B'):
            data_vars = await self.get_queue_data(comm)
            data_names = ['temp_calib_slope', 'temp_calib_intercept']
            return dict(zip(data_names, data_vars))


    async def get_eeprom_firmware(self):
        comm = 'EEPROM_FIRMWARE'
        if self.snd_rava_msg(comm):
            version_major, version_minor, version_patch, modules = await self.get_queue_data(comm)

            version = '{}.{}.{}'.format(version_major, version_minor, version_patch)
            health_startup_enabled = modules & 1 << 0 != 0
            health_continuous_enabled = modules & 1 << 1 != 0
            led_enabled = modules & 1 << 2 != 0
            lamp_enabled = modules & 1 << 3 != 0
            peripherals_enabled = modules & 1 << 4 != 0
            serial1_enabled = modules & 1 << 5 != 0
            return {'version':version,
                    'health_startup_enabled':health_startup_enabled,
                    'health_continuous_enabled':health_continuous_enabled,
                    'led_enabled':led_enabled,
                    'lamp_enabled':lamp_enabled,
                    'peripherals_enabled':peripherals_enabled,
                    'serial1_enabled':serial1_enabled
                    }


    async def get_eeprom_pwm(self):
        comm = 'EEPROM_PWM'
        rava_send = True
        if self.snd_rava_msg(comm, [rava_send], 'B'):
            freq_id, duty = await self.get_queue_data(comm)
            return {'freq_id':freq_id, 'freq_str':D_PWM_FREQ_INV[freq_id], 'duty':duty}


    async def get_eeprom_rng(self):
        comm = 'EEPROM_RNG'
        rava_send = True
        if self.snd_rava_msg(comm, [rava_send], 'B'):
            sampling_interval_us = await self.get_queue_data(comm)
            return {'sampling_interval_us':sampling_interval_us}


    async def get_eeprom_led(self):
        comm = 'EEPROM_LED'
        rava_send = True
        if self.snd_rava_msg(comm, [rava_send], 'B'):
            led_attached = await self.get_queue_data(comm)
            return {'led_attached':led_attached}


    async def get_eeprom_lamp(self):
        comm = 'EEPROM_LAMP'
        rava_send = True
        if self.snd_rava_msg(comm, [rava_send], 'B'):
            data_vars = await self.get_queue_data(comm)
            data_names = ['exp_dur_max_ms', 'exp_z_significant', 'exp_mag_smooth_n_trials']
            return dict(zip(data_names, data_vars))


    #####################
    ## PWM

    async def get_pwm_setup(self):
        rava_send = True
        comm = 'PWM_SETUP'
        if self.snd_rava_msg(comm, [rava_send], 'B'):
            freq_id, duty = await self.get_queue_data(comm)
            return {'freq_id':freq_id, 'freq_str':D_PWM_FREQ_INV[freq_id], 'duty':duty}


    #####################
    ## RNG

    async def get_rng_setup(self):
        comm = 'RNG_SETUP'
        rava_send = True
        if self.snd_rava_msg(comm, [rava_send], 'B'):
            sampling_interval_us = await self.get_queue_data(comm)
            return {'sampling_interval_us':sampling_interval_us}


    async def get_rng_pulse_counts(self, n_counts):
        comm = 'RNG_PULSE_COUNTS'
        if self.snd_rava_msg(comm, [n_counts], 'L'):
            counts_bytes_a, counts_bytes_b = await self.get_queue_data(comm)

            counts_a = bytes_to_numlist(counts_bytes_a, 'B')
            counts_b = bytes_to_numlist(counts_bytes_b, 'B')
            return counts_a, counts_b


    async def get_rng_bits(self, bit_source_id):
        if bit_source_id not in D_RNG_BIT_SRC_INV:
            lg.error('{} RNG Bits: Unknown bit_source_id {}'.format(self.dev_name, bit_source_id))
            return None

        comm = 'RNG_BITS'
        if self.snd_rava_msg(comm, [bit_source_id], 'B'):
            bit_source_id_recv, *bits = await self.get_queue_data(comm)

            if bit_source_id == D_RNG_BIT_SRC['AB_RND']:
                bit_type_str = D_RNG_BIT_SRC_INV[bit_source_id_recv]
                return [bit_type_str, bits[0]]
            elif bit_source_id == D_RNG_BIT_SRC['AB']:
                return list(bits)
            else:
                return bits[0]


    async def get_rng_bytes(self, n_bytes, postproc_id=D_RNG_POSTPROC['NONE'], request_id=0, list_output=True):
        if postproc_id not in D_RNG_POSTPROC_INV:
            lg.error('{} RNG Bytes: Unknown postproc_id {}'.format(self.dev_name, postproc_id))
            return None

        comm = 'RNG_BYTES'
        if self.snd_rava_msg(comm, [n_bytes, postproc_id, request_id], 'LBB'):
            bytes_data = await self.get_queue_data(comm, comm_ext_id=request_id)

            if bytes_data is not None:
                rng_bytes_a, rng_bytes_b = bytes_data

                if list_output:
                    rng_a = bytes_to_numlist(rng_bytes_a, 'B')
                    rng_b = bytes_to_numlist(rng_bytes_b, 'B')
                    return rng_a, rng_b
                else:
                    return rng_bytes_a, rng_bytes_b

            else:
                return None, None


    async def get_rng_int8s(self, n_ints, int_delta):
        if n_ints >= 2**32:
            lg.error('{} RNG Ints: Provide n_ints as a 32-bit integer'.format(self.dev_name))
            return None
        if int_delta >= 2**8:
            lg.error('{} RNG Ints: Provide int_delta as a 8-bit integer'.format(self.dev_name))
            return None
        if int_delta < 2:
            lg.error('{} RNG Ints: Provide int_delta > 1'.format(self.dev_name))
            return None

        comm = 'RNG_INT8S'
        if self.snd_rava_msg(comm, [n_ints, int_delta], 'LB'):
            ints_bytes = await self.get_queue_data(comm)
            return bytes_to_numlist(ints_bytes, 'B')


    async def get_rng_int16s(self, n_ints, int_delta):
        if n_ints >= 2**32:
            lg.error('{} RNG Ints: Provide n_ints as a 32-bit integer'.format(self.dev_name))
            return None
        if int_delta >= 2**16:
            lg.error('{} RNG Ints: Provide int_delta as a 16-bit integer'.format(self.dev_name))
            return None
        if int_delta < 2:
            lg.error('{} RNG Ints: Provide int_delta > 1'.format(self.dev_name))
            return None

        comm = 'RNG_INT16S'
        if self.snd_rava_msg(comm, [n_ints, int_delta], 'LH'):
            ints_bytes = await self.get_queue_data(comm)
            return bytes_to_numlist(ints_bytes, 'H')


    async def get_rng_floats(self, n_floats):
        if n_floats >= 2**32:
            lg.error('{} RNG Floats: Provide n_floats as a 32-bit integer'.format(self.dev_name))
            return None

        comm = 'RNG_FLOATS'
        if self.snd_rava_msg(comm, [n_floats], 'L'):
            ints_bytes = await self.get_queue_data(comm)
            return bytes_to_numlist(ints_bytes, 'f')


    async def get_rng_byte_stream_data(self, list_output=True):
        if not self.rng_streaming:
            lg.warning('{} RNG Stream: Streaming is disabled'.format(self.dev_name))
            return None, None

        comm = 'RNG_STREAM_BYTES'
        bytes_data = await self.get_queue_data(comm)

        # Timeout?
        if bytes_data is not None:
            rng_bytes_a, rng_bytes_b = bytes_data

            if list_output:
                rng_a = bytes_to_numlist(rng_bytes_a, 'B')
                rng_b = bytes_to_numlist(rng_bytes_b, 'B')
                return rng_a, rng_b
            else:
                return rng_bytes_a, rng_bytes_b

        else:
            return None, None


    async def get_rng_byte_stream_status(self):
        comm = 'RNG_STREAM_STATUS'
        if self.snd_rava_msg(comm):
            return await self.get_queue_data(comm)


    #####################
    ## HEALTH

    async def get_health_startup_results(self):
        comm = 'HEALTH_STARTUP_RESULTS'
        if self.snd_rava_msg(comm):
            success, pc_avg_vars, pc_avg_diff_vars, bias_bit_vars, bias_byte_vars = await self.get_queue_data(comm)
            data_vars = [pc_avg_vars, pc_avg_diff_vars, bias_bit_vars, bias_byte_vars]
            data_names = ['pulse_count', 'pulse_count_diff', 'bit_bias', 'byte_bias']
            return success, dict(zip(data_names, data_vars))


    async def get_health_continuous_errors(self):
        comm = 'HEALTH_CONTINUOUS_ERRORS'
        if self.snd_rava_msg(comm):
            data_vars = await self.get_queue_data(comm)
            n_errors = sum(data_vars)
            data_names = ['repetitive_count_a', 'repetitive_count_b',
                          'adaptative_proportion_a', 'adaptative_proportion_b']
            return n_errors, dict(zip(data_names, data_vars))


    #####################
    ## PERIPHERALS

    async def get_periph_digi_state(self, periph_id=1):
        if periph_id == 0 or periph_id > 5:
            lg.error('{} Periph: Provide a periph_id between 1 and 5'.format(self.dev_name))
            return None

        comm = 'PERIPH_READ'
        if self.snd_rava_msg(comm, [periph_id], 'B'):
            digi_state, digi_mode = await self.get_queue_data(comm)
            return digi_state, D_PERIPH_MODES_INV[digi_mode]


    async def get_periph_d2_input_capture(self):
        comm = 'PERIPH_D2_TIMER3_INPUT_CAPTURE'
        return await self.get_queue_data(comm)


    async def get_periph_d5_adc_read(self, ref_5v=1, clk_prescaler=0, oversampling_n_bits=0, on=True):
        if clk_prescaler == 0 or clk_prescaler > 7:
            lg.error('{} Periph D5: Provide a clk_prescaler between 1 and 7'.format(self.dev_name))
            return None
        if oversampling_n_bits > 6:
            lg.error('{} Periph D5: Provide a oversampling_n_bits <= 6'.format(self.dev_name))
            return None

        comm = 'PERIPH_D5_ADC'
        if self.snd_rava_msg(comm, [on, ref_5v, clk_prescaler, oversampling_n_bits], 'BBBB'):
            if on:
                return await self.get_queue_data(comm)


    #####################
    ## INTERFACES

    async def get_interface_ds18bs0(self):
        comm = 'INTERFACE_DS18B20'
        if self.snd_rava_msg(comm):
            return await self.get_queue_data(comm)